tests/net/lib/http_server/prototype/src/main.c

```diff
@@ -0,0 +1,466 @@
+/*
+ * Copyright (c) 2023, Emna Rekik
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "server_functions.h"
+
+#include <string.h>
+
+#include <zephyr/net/http/service.h>
+#include <zephyr/net/socket.h>
+#include <zephyr/posix/sys/eventfd.h>
+#include <zephyr/ztest.h>
+
+#define SUPPORT_BACKWARD_COMPATIBILITY 1
+#define SUPPORT_HTTP_SERVER_UPGRADE    2
+#define BUFFER_SIZE                    256
+#define STACK_SIZE                     8192
+#define MY_IPV4_ADDR                   "127.0.0.1"
+#define SERVER_PORT                    8080
+#define TIMEOUT                        1000
+
+static struct k_sem server_sem;
+
+static K_THREAD_STACK_DEFINE(server_stack, STACK_SIZE);
+
+static struct k_thread server_thread;
+
+/* Magic, SETTINGS[0], HEADERS[1]: GET /, HEADERS[3]: GET /index.html */
+static const unsigned char Frame1[] = {
+	0x50, 0x52, 0x49, 0x20, 0x2a, 0x20, 0x48, 0x54, 0x54, 0x50, 0x2f, 0x32, 0x2e, 0x30, 0x0d,
+	0x0a, 0x0d, 0x0a, 0x53, 0x4d, 0x0d, 0x0a, 0x0d, 0x0a, 0x00, 0x00, 0x0c, 0x04, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x03, 0x00, 0x00, 0x00, 0x64, 0x00, 0x04, 0x00, 0x00, 0xff, 0xff,
+	0x00, 0x00, 0x21, 0x01, 0x05, 0x00, 0x00, 0x00, 0x01, 0x82, 0x84, 0x86, 0x41, 0x8a, 0x0b,
+	0xe2, 0x5c, 0x0b, 0x89, 0x70, 0xdc, 0x78, 0x0f, 0x03, 0x53, 0x03, 0x2a, 0x2f, 0x2a, 0x90,
+	0x7a, 0x8a, 0xaa, 0x69, 0xd2, 0x9a, 0xc4, 0xc0, 0x57, 0x68, 0x0b, 0x83, 0x00, 0x00, 0x07,
+	0x01, 0x05, 0x00, 0x00, 0x00, 0x03, 0x82, 0x85, 0x86, 0xc0, 0xbf, 0x90, 0xbe};
+
+/* SETTINGS[0] */
+static const unsigned char Frame2[] = {0x00, 0x00, 0x00, 0x04, 0x01, 0x00, 0x00, 0x00, 0x00};
+
+/* GOAWAY[0] */
+static const unsigned char Frame3[] = {0x00, 0x00, 0x08, 0x07, 0x00, 0x00, 0x00, 0x00, 0x00,
+				       0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+static uint16_t test_http_service_port = htons(SERVER_PORT);
+HTTP_SERVICE_DEFINE(test_http_service, CONFIG_NET_CONFIG_MY_IPV4_ADDR, &test_http_service_port, 1,
+		    10, NULL);
+
+static const char index_html_gz[] = "Hello, World!";
+struct http_resource_detail_static index_html_gz_resource_detail = {
+	.common = {
+			.type = HTTP_RESOURCE_TYPE_STATIC,
+		},
+	.static_data = index_html_gz,
+	.static_data_len = sizeof(index_html_gz),
+};
+
+HTTP_RESOURCE_DEFINE(index_html_gz_resource, test_http_service, "/",
+		     &index_html_gz_resource_detail);
+
+static void server_thread_fn(void *arg0, void *arg1, void *arg2)
```

Comment:
the server thread should really be an implementation detail in `subsys/net/.../http_server.c`